File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool sameTree(TreeNode* root1, TreeNode* root2) {
        if (!root1 && !root2) return true;

        if (root1 && root2 && root1->val == root2->val) {
            return  sameTree(root1->left, root2->left) &&
                    sameTree(root1->right, root2->right);
        }
        return false;
    }

    bool isSubtree(TreeNode* root, TreeNode* subRoot) {
        if (!subRoot) return true;
        if (!root) return false;
        if (sameTree(root, subRoot)) return true;
        return  isSubtree(root->left, subRoot) ||
                isSubtree(root->right, subRoot);
    }
};

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        self.res = float("-inf")

        def dfs(node):
            if not node:
                return 0
            
            dfs(node.left)
            dfs(node.right)
            left = node.left.val if node.left else 0
            right = node.right.val if node.right else 0
            self.res = max(self.res, node.val, node.val + left, node.val + right, node.val + left + right)
            node.val = max(node.val, node.val + left, node.val + right)

        dfs(root)
        return self.res